### PR TITLE
api: Add pause or resume checker REST api

### PR DESCRIPTION
**Pull request by peng1999**
Signed-off-by: Peng Guanwen <pg999w@outlook.com>

### What problem does this PR solve?

When TiKV is in import mode, we don't want PD to merge regions. This PR add a new API to stop region merge for a given time.

### What is changed and how it works?

A new PD REST API `/checker/{name}` with a paramenter `{delay}` is added. When called, PD will not trigger corresponding checker in the following `{delay}` seconds.

### Check List

- Unit test
- Integration test

Code changes

- Has HTTP API interfaces change (Don't forget to [add the declarative for API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))

### Release note

```release-note
Add a `/checker/{name}` API to pause region merge.
```

### CODE REVIEW

**Comment by nolouch** [comment]
Can it be more general, like schedulers, I can pause different schedulers. In here, may pause different checker.

**Comment by Yisaer** [comment]
> Can it be more general, like schedulers, I can pause different schedulers. In here, may pause different checker.

+1

**Comment by peng1999** [comment]
@Yisaer I've changed the API to `/checker/{name}`, so that we can stop different checkers.

File: server/api/scheduler.go [hunk 1]
```diff
@@ -324,6 +324,36 @@ func (h *schedulerHandler) PauseOrResume(w http.ResponseWriter, r *http.Request)
...
+// @Summary Pause or resume region merge.
+// @Accept json
+// @Param name path string true "The name of the scheduler."
+// @Param body body object true "json params"
+// @Produce json
+// @Success 200 {string} string "Pause or resume the scheduler successfully."
+// @Failure 400 {string} string "Bad format request."
+// @Failure 500 {string} string "PD server failed to proceed the request."
+// @Router /checker/{name} [post]
+func (h *schedulerHandler) PauseOrResumeChecker(w http.ResponseWriter, r *http.Request) {
```
Comment (Yisaer) [inline server/api/scheduler.go:338]: I think using `checkerHandler` will be better

File: server/schedule/checker/checker_pause.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package checker
```
Comment (Yisaer) [inline server/schedule/checker/checker_pause.go:1]: add license

**Review by Yisaer — COMMENTED** [review]
Could you add unit test?

File: server/schedule/checker/checker_pause.go (hunk 2 above)
Comment (peng1999) [reply server/schedule/checker/checker_pause.go:15]: License added.

File: server/api/scheduler.go (hunk 1 above)
Comment (peng1999) [reply server/api/scheduler.go]: Function renamed.

File: server/api/scheduler.go [hunk 3]
```diff
@@ -324,6 +324,36 @@ func (h *schedulerHandler) PauseOrResume(w http.ResponseWriter, r *http.Request)
...
+// @Summary Pause or resume region merge.
+// @Accept json
+// @Param name path string true "The name of the scheduler."
+// @Param body body object true "json params"
+// @Produce json
+// @Success 200 {string} string "Pause or resume the scheduler successfully."
+// @Failure 400 {string} string "Bad format request."
+// @Failure 500 {string} string "PD server failed to proceed the request."
+// @Router /checker/{name} [post]
+func (h *schedulerHandler) CheckerHandler(w http.ResponseWriter, r *http.Request) {
```
Comment (rleungx) [inline server/api/scheduler.go:338]: Why is the checker's code in scheduler.go?

**Comment by peng1999** [comment]
> Could you add unit test?

Which code should be test?

File: server/api/scheduler.go (hunk 3 above)
Comment (peng1999) [reply server/api/scheduler.go]: Sorry. It's my mistake. The function is now moved to a `checker.go` file.

**Review by Yisaer — COMMENTED** [review]
you can add testcase like `testScheduleSuite.testPauseOrResume`

File: server/api/checker.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+package api
+
+import (
+	"github.com/gorilla/mux"
+	"github.com/tikv/pd/pkg/apiutil"
+	"github.com/tikv/pd/server"
+	"github.com/unrolled/render"
+	"net/http"
+)
```
[the remark this answers is not in the document]
Comment (peng1999) [reply server/api/checker.go]: The imports are grouped.

File: server/schedule/checker/joint_state_checker.go [hunk 5]
```diff
@@ -37,6 +38,10 @@ func NewJointStateChecker(cluster opt.Cluster) *JointStateChecker {
 // Check verifies a region's role, creating an Operator if need.
 func (c *JointStateChecker) Check(region *core.RegionInfo) *operator.Operator {
 	checkerCounter.WithLabelValues("joint_state_checker", "check").Inc()
+	if c.IsPaused() {
+		checkerCounter.WithLabelValues("joint_state_checker", "skipped").Inc()
```
Comment (nolouch) [inline server/schedule/checker/joint_state_checker.go:42]:
```suggestion
		checkerCounter.WithLabelValues("joint_state_checker", "paused").Inc()
```
it's more intuitive.

File: server/schedule/checker/learner_checker.go [hunk 6]
```diff
@@ -36,6 +37,10 @@ func NewLearnerChecker(cluster opt.Cluster) *LearnerChecker {
 
 // Check verifies a region's role, creating an Operator if need.
 func (l *LearnerChecker) Check(region *core.RegionInfo) *operator.Operator {
+	if l.IsPaused() {
+		checkerCounter.WithLabelValues("learner_checker", "skipped").Inc()
```
Comment (nolouch) [inline server/schedule/checker/learner_checker.go:41]: same as above.

File: server/api/checker_test.go [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+	//handler := s.svr.GetHandler()
+
+	// test pause.
+	input := make(map[string]interface{})
+	input["delay"] = 30
+	pauseArgs, err := json.Marshal(input)
+	c.Assert(err, IsNil)
+	err = postJSON(testDialClient, s.urlPrefix+"/"+name, pauseArgs)
+	c.Assert(err, IsNil)
+	//isPaused, err := handler.IsCheckerPaused(createdName)
```
Comment (nolouch) [inline server/api/checker_test.go:92]: not implement?

**Review by nolouch — COMMENTED** [review]
rest lgtm.

File: server/api/checker_test.go (hunk 7 above)
Comment (peng1999) [reply server/api/checker_test.go:92]: No, `IsCheckerPaused` is not implemented. This PR is mainly for https://github.com/tikv/client-java/issues/268, and currently no one needs to check if a checker is paused. But I left the comments for future convenience. Once `IsCheckerPaused` is implemented, we can add these lines.

File: server/schedule/checker/joint_state_checker.go (hunk 5 above)
Comment (peng1999) [reply server/schedule/checker/joint_state_checker.go]: done.

File: server/schedule/checker/learner_checker.go (hunk 6 above)
Comment (peng1999) [reply server/schedule/checker/learner_checker.go]: done.

File: server/api/checker_test.go (hunk 7 above)
Comment (Yisaer) [reply server/api/checker_test.go:92]: I think you can implement `IsCheckerPaused` and make this testcase completed. The rest LGTM.

Comment (peng1999) [reply server/api/checker_test.go]: Done.

**Comment by rleungx** [comment]
How do we know if a checker is paused?

**Comment by peng1999** [comment]
> How do we know if a checker is paused?

If `checker.PauseController.delayUntil` is greater than current time, we say the checker is paused. See below code:

https://github.com/tikv/pd/blob/fdfcfbda526c5548b024c16d5e66400d7b2ac1f8/server/schedule/checker/checker_pause.go#L28-L31

**Comment by rleungx** [comment]
> .

Actually, what I want to emphasize is if we use a long time to delay a checker and there is no way to let us know this checker is paused.

**Comment by peng1999** [comment]
> Actually, what I want to emphasize is if we use a long time to delay a checker and there is no way to let us know this checker is paused.

Ok. I'll add a REST API to check if a checker is paused.

**Review by Yisaer — APPROVED** [review]
LGTM

**Comment by peng1999** [comment]
@rleungx @nolouch It seems that all issues are resolved. Could you please review this PR again?

File: server/api/checker.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+	}
+}
+
+// FIXME: details of input json body params
+// @Tags checker
+// @Summary Get if checker is paused
+// @Param name path string true "The name of the scheduler."
+// @Produce json
+// @Success 200 {string} string "Pause or resume the scheduler successfully."
+// @Failure 400 {string} string "Bad format request."
```
Comment (rleungx) [inline server/api/checker.go:78]: Where is 400 error?

File: server/api/checker_test.go [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	. "github.com/pingcap/check"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/tikv/pd/server"
+	_ "github.com/tikv/pd/server/schedulers"
```
Comment (rleungx) [inline server/api/checker_test.go:25]: Do we need it?

File: server/handler.go [hunk 10]
```diff
@@ -242,6 +251,24 @@ func (h *Handler) PauseOrResumeScheduler(name string, t int64) error {
 	return err
 }
 
+// PauseOrResumeChecker pauses checker for delay seconds or resume checker
+// t == 0 : resume checker.
+// t > 0 : checker delays t seconds.
```
Comment (rleungx) [inline server/handler.go:256]: What if t < 0?

**Review by nolouch — APPROVED** [review]

File: server/api/checker.go (hunk 8 above)
Comment (peng1999) [reply server/api/checker.go]: Fixed.

File: server/api/checker_test.go (hunk 9 above)
Comment (peng1999) [reply server/api/checker_test.go]: No, we needn't. Deleted.

File: server/handler.go (hunk 10 above)
Comment (peng1999) [reply server/handler.go:256]: t < 0 will behave the same as t == 0, i.e. resume the checker, if we do not consider `int64` overflow. But passing a argument t < 0 should not encouraged. Should this be documented?

Comment (rleungx) [reply server/handler.go:256]: Yes, or return a request error?

**Comment by marsishandsome** [comment]
run `mvn clean package -DskipTests` to format the code

File: server/handler.go (hunk 10 above)
Comment (peng1999) [reply server/handler.go:256]: Fixed. A negative delay now gives a 400 status.

File: server/api/checker_test.go [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+	err = postJSON(testDialClient, s.urlPrefix+"/"+name, pauseArgs)
+	c.Assert(err, NotNil)
+	input["delay"] = 30
+	pauseArgs, err = json.Marshal(input)
+	c.Assert(err, IsNil)
+	err = postJSON(testDialClient, s.urlPrefix+"/"+name, pauseArgs)
+	c.Assert(err, IsNil)
+	isPaused, err := handler.IsCheckerPaused(name)
+	c.Assert(err, IsNil)
+	c.Assert(isPaused, Equals, true)
```
Comment (rleungx) [inline server/api/checker_test.go:111]: We can also use IsTrue and IsFalse

Comment (peng1999) [reply server/api/checker_test.go]: Fixed.

**Review by rleungx — COMMENTED** [review]
LGTM and this PR has enough approval.

**Pull request merged**